File: modules/openlmis-web/src/test/java/org/openlmis/web/controller/SupervisoryNodeControllerTest.java
/*
 * This program is part of the OpenLMIS logistics management information system platform software.
 * Copyright © 2013 VillageReach
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
 * You should have received a copy of the GNU Affero General Public License along with this program.  If not, see http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org. 
 */

package org.openlmis.web.controller;

import org.junit.Before;
import org.junit.Test;
import org.junit.experimental.categories.Category;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;
import org.openlmis.core.domain.SupervisoryNode;
import org.openlmis.core.exception.DataException;
import org.openlmis.core.service.MessageService;
import org.openlmis.core.service.SupervisoryNodeService;
import org.openlmis.db.categories.UnitTests;
import org.openlmis.core.web.OpenLmisResponse;
import org.springframework.http.ResponseEntity;
import org.springframework.mock.web.MockHttpServletRequest;

import java.util.ArrayList;
import java.util.List;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.*;
import static org.openlmis.web.controller.SupervisoryNodeController.SUPERVISORY_NODES;

@Category(UnitTests.class)
@RunWith(MockitoJUnitRunner.class)
public class SupervisoryNodeControllerTest {

  @Mock
  private SupervisoryNodeService supervisoryNodeService;

  @Mock
  MessageService messageService;

  @InjectMocks
  private SupervisoryNodeController controller;

  public static final String USER_ID = "USER_ID";
  private MockHttpServletRequest request = new MockHttpServletRequest();

  @Before
  public void setUp() {
    setLoggedInUserId();
  }

  @Test
  public void shouldGetAllSupervisoryNodes() {
    List<SupervisoryNode> expectedSupervisoryNodes = new ArrayList<>();
    when(supervisoryNodeService.getAll()).thenReturn(expectedSupervisoryNodes);

    ResponseEntity<OpenLmisResponse> response = controller.getAll();

    verify(supervisoryNodeService).getAll();
    List<SupervisoryNode> actual = (List<SupervisoryNode>) response.getBody().getData().get(SUPERVISORY_NODES);
    assertThat(actual, is(expectedSupervisoryNodes));
  }

  @Test
  public void shouldGetSupervisoryNodeById() {
    Long id = 1L;
    SupervisoryNode expectedNode = new SupervisoryNode();
    when(supervisoryNodeService.getSupervisoryNode(id)).thenReturn(expectedNode);

    SupervisoryNode actualNode = controller.getById(id);

    verify(supervisoryNodeService).getSupervisoryNode(id);
    assertThat(actualNode, is(expectedNode));
  }

  @Test
  public void shouldGetFilteredNodesForRequestedQuery() {
    String query = "Node1";
    List<SupervisoryNode> supervisoryNodes = new ArrayList<>();
    when(supervisoryNodeService.getFilteredSupervisoryNodesByName(query)).thenReturn(supervisoryNodes);

    List<SupervisoryNode> filteredNodes = controller.getFilteredNodes(query);

    verify(supervisoryNodeService).getFilteredSupervisoryNodesByName(query);
    assertThat(filteredNodes, is(supervisoryNodes));
  }

  @Test
  public void shouldInsertSupervisoryNodeSuccessfully() {
    SupervisoryNode supervisoryNode = new SupervisoryNode(1L);
    supervisoryNode.setName("Node 1");
    when(messageService.message("message.supervisory.node.created.success", supervisoryNode.getName())).thenReturn("success");

    ResponseEntity<OpenLmisResponse> responseEntity = controller.insert(supervisoryNode, request);

    verify(supervisoryNodeService).save(supervisoryNode);
    assertThat((Long) responseEntity.getBody().getData().get("supervisoryNodeId"), is(supervisoryNode.getId()));
    assertThat(responseEntity.getBody().getSuccessMsg(), is("success"));
    assertThat(supervisoryNode.getCreatedBy(), is(1L));
    assertThat(supervisoryNode.getModifiedBy(), is(1L));
  }

  @Test
  public void shouldReturnErrorMessageWhenExceptionOccursOnSupervisoryNodeInsert() {
    SupervisoryNode supervisoryNode = new SupervisoryNode();
    doThrow(new DataException("error")).when(supervisoryNodeService).save(supervisoryNode);

    ResponseEntity<OpenLmisResponse> responseEntity = controller.insert(supervisoryNode, request);

    verify(supervisoryNodeService).save(supervisoryNode);
    verify(messageService, never()).message(anyString(), anyString());
    assertThat(responseEntity.getBody().getErrorMsg(), is("error"));
    assertThat(supervisoryNode.getCreatedBy(), is(1L));
    assertThat(supervisoryNode.getModifiedBy(), is(1L));
  }

  @Test
  public void shouldUpdateSupervisoryNodeSuccessfully() {
    SupervisoryNode supervisoryNode = new SupervisoryNode(1L);
    supervisoryNode.setName("Node 1");
    when(messageService.message("message.supervisory.node.updated.success", supervisoryNode.getName())).thenReturn(
      "success");

    ResponseEntity<OpenLmisResponse> responseEntity = controller.update(supervisoryNode, supervisoryNode.getId(),
      request);

    verify(supervisoryNodeService).save(supervisoryNode);
    assertThat((Long) responseEntity.getBody().getData().get("supervisoryNodeId"), is(supervisoryNode.getId()));
    assertThat(responseEntity.getBody().getSuccessMsg(), is("success"));
    assertThat(supervisoryNode.getModifiedBy(), is(1L));
  }

  @Test
  public void shouldReturnErrorMessageWhenExceptionOccursOnSupervisoryNodeUpdate() {
    SupervisoryNode supervisoryNode = new SupervisoryNode(1L);
    doThrow(new DataException("error")).when(supervisoryNodeService).save(supervisoryNode);

    ResponseEntity<OpenLmisResponse> responseEntity = controller.update(supervisoryNode, 1L, request);

    verify(supervisoryNodeService).save(supervisoryNode);
    verify(messageService, never()).message(anyString(), anyString());
    assertThat(responseEntity.getBody().getErrorMsg(), is("error"));
    assertThat(supervisoryNode.getModifiedBy(), is(1L));
  }

  @Test
  public void shouldGetTopLevelNodesForRequestedQuery() {
    String query = "Node1";
    List<SupervisoryNode> supervisoryNodes = new ArrayList<>();
    when(supervisoryNodeService.searchTopLevelSupervisoryNodesByName(query)).thenReturn(supervisoryNodes);

    List<SupervisoryNode> filteredNodes = controller.searchTopLevelSupervisoryNodesByName(query);

    verify(supervisoryNodeService).searchTopLevelSupervisoryNodesByName(query);
    assertThat(filteredNodes, is(supervisoryNodes));
  }

  private void setLoggedInUserId() {
    request.getSession().setAttribute(USER_ID, 1L);
  }
}

File: modules/openlmis-web/src/test/javascript/specs/supervisory-node/controller/supervisory-node-controller-test.js
/*
 * This program is part of the OpenLMIS logistics management information system platform software.
 * Copyright © 2013 VillageReach
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
 * You should have received a copy of the GNU Affero General Public License along with this program.  If not, see http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org.
 */

describe("Supervisory Node Controller", function () {

  var scope, ctrl, supervisoryNode, element, $httpBackend, location;
  beforeEach(module('openlmis'));

  beforeEach(inject(function ($rootScope, _$httpBackend_, $controller, $location) {
    scope = $rootScope.$new();
    $httpBackend = _$httpBackend_;
    location = $location;
    scope.query = "Nod";
    supervisoryNode = {code: "N1", name: "Node 1"};
    ctrl = $controller('SupervisoryNodeController', {$scope: scope, supervisoryNode: supervisoryNode});
  }));

  it('should set supervisory node in scope', function () {
    expect(scope.supervisoryNode).toBe(supervisoryNode);
  });

  it('should get all parent nodes in scope if query length is greater than 3 in case of add', function () {
    scope.query = "Nod";
    var node1 = {"id": 1, "code": "N1", "name": "Node 1"};
    var node2 = {"id": 2, "code": "N2", "name": "Node 2"};
    var node3 = {"id": 3, "code": "N3", "name": "Node 3"};
    var response = {"supervisoryNodeList": [node1, node2, node3]};

    $httpBackend.when('GET', '/search-supervisory-nodes.json?searchParam=' + scope.query).respond(response);
    scope.showParentNodeSearchResults();
    $httpBackend.flush();

    expect(scope.parentNodes).toEqual([node1, node2, node3]);
    expect(scope.filteredNodes).toEqual([node1, node2, node3]);
    expect(scope.previousQuery).toEqual("Nod");
    expect(scope.nodeResultCount).toEqual(3);
  });

  it('should do client side filtering when previous query and current query are same', function () {
    spyOn($httpBackend, "expectGET");
    scope.query = "Node";
    scope.previousQuery = "Nod";
    var node2 = {"id": 2, "code": "N2", "name": "Node 2"};
    var node3 = {"id": 3, "code": "N3", "name": "Node 3"};
    var node4 = {"id": 3, "code": "N3", "name": "Nodde 4"};
    scope.parentNodes = [node2, node3, node4];

    scope.showParentNodeSearchResults();

    expect($httpBackend.expectGET).not.toHaveBeenCalledWith('/search-supervisory-nodes.json?searchParam=' + scope.query);
    expect(scope.filteredNodes).toEqual([node2, node3]);
    expect(scope.previousQuery).toEqual("Nod");
    expect(scope.nodeResultCount).toEqual(2);
  });

  it('should not search results if query is undefined', function () {
    spyOn($httpBackend, "expectGET");
    scope.query = undefined;

    scope.showParentNodeSearchResults();

    expect($httpBackend.expectGET).not.toHaveBeenCalledWith('/search-supervisory-nodes.json?searchParam=' + scope.query);
  });

  it('should not search results if query length is less than 3', function () {
    spyOn($httpBackend, "expectGET");
    scope.query = "No";

    scope.showParentNodeSearchResults();

    expect($httpBackend.expectGET).not.toHaveBeenCalledWith('/search-supervisory-nodes.json?searchParam=' + scope.query);
  });

  it('should select parent node and clear scope fields', function () {
    var node = {code: "SN2", name: "Node 2"};
    scope.setSelectedParentNode(node);
    expect(scope.supervisoryNode.parent).toBe(node);
    expect(scope.query).toBeUndefined();
    expect(scope.nodeResultCount).toBeUndefined();
    expect(scope.filteredFacilities).toBeUndefined();
    expect(scope.previousQuery).toBeUndefined();
  });

  it('should delete parent node', function () {
    scope.supervisoryNode.parent = {"Code": "N1", "Name": "Node 1"};

    scope.deleteParentNode();

    expect(scope.supervisoryNode.parent).toBeUndefined();
  });

  it('should take to search page on cancel', function () {
    scope.cancel();
    expect(scope.$parent.message).toEqual("");
    expect(scope.$parent.supervisoryNodeId).toBeUndefined();
    expect(location.path()).toEqual('/#/search');
  });

  it('should not save supervisory node if invalid', function () {
    scope.supervisoryNodeForm = {"$error": {"required": true}};

    scope.save();

    expect(scope.error).toEqual("form.error");
    expect(scope.showError).toBeTruthy();
  });

  it('should not save supervisory node if associated facility missing', function () {
    scope.supervisoryNodeForm = {"$error": {"required": false}};
    scope.supervisoryNode = {facility: undefined};

    scope.save();

    expect(scope.error).toEqual("form.error");
    expect(scope.showError).toBeTruthy();
  });

  it('should insert supervisory node', function () {
    scope.supervisoryNode = {"code": "N100", "name": "node 100", "facility": {"code": "F10", "name": "village dispensary"}};
    scope.supervisoryNodeForm = {"$error": {"required": false}};

    $httpBackend.expectPOST('/supervisory-nodes.json', scope.supervisoryNode).respond(200, {"success": "Saved successfully", "supervisoryNode": scope.supervisoryNode});
    scope.save();
    $httpBackend.flush();

    expect(scope.error).toEqual("");
    expect(scope.showError).toBeFalsy();
    expect(scope.$parent.message).toEqual("Saved successfully");
    expect(scope.$parent.supervisoryNodeId).toEqual(supervisoryNode.id);
  });

  it('should update supervisory node', function () {
    scope.supervisoryNodeForm = {"$error": {"required": false}};
    supervisoryNode.id = 4;
    supervisoryNode.facility = {"code": "F10", "name": "village dispensary"};

    $httpBackend.expectPUT('/supervisory-nodes/' + supervisoryNode.id + '.json', supervisoryNode).respond(200, {"success": "Saved successfully", "supervisoryNodeId": supervisoryNode.id});
    scope.save();
    $httpBackend.flush();

    expect(scope.error).toEqual("");
    expect(scope.showError).toBeFalsy();
    expect(scope.$parent.message).toEqual("Saved successfully");
    expect(scope.$parent.supervisoryNodeId).toEqual(supervisoryNode.id);
  });

  it('should toggle slider', function () {
    scope.showSlider = true;

    scope.toggleSlider();

    expect(scope.showSlider).toBeFalsy();
    expect(scope.extraParams.virtualFacility).toBeNull();
    expect(scope.extraParams.enabled).toBeTruthy();
  });

});
